clientv3/watch.go

```diff
@@ -614,13 +603,34 @@ func (w *watchGrpcStream) serveSubstream(ws *watcherStream, resumec chan struct{
 				// shutdown from closeSubstream
 				return
 			}
-			// TODO pause channel if buffer gets too large
-			ws.buf = append(ws.buf, wr)
+
+			if wr.Created {
+				select {
+				case ws.initReq.retc <- ws.outc:
+					// send first creation event only if requested
+					if ws.initReq.createdNotify {
+						ws.outc <- *wr
```

Comment:
Realized what's going wrong here. The channel is buffered so the first read will clear the buffer slot on `retc`, causing the next write to take the slot and run the `<- *wr` path a second time. I think the fix is:

```go
if ws.initReqc.retc != nil {
    ws.initReq.retc <- ws.outc
    ws.initReq.retc = nil
    if ws.initReq.createNotify { ws.outc <- *wr }
}
```